python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py

```diff
@@ -17,227 +17,213 @@
+        self.VarQuantScales = {}
         self.max_range = {}
-        self.new_output = {}
         self.s8_max = 127
 
     def apply(self, graph):
-        """
-        Quantize the graph for running MKL-DNN INT8 inference. According 
-        to activation quantization type, the graph will transform fake 
-        quantize ops to quantize ops and remove the fake dequantize ops.
-      
-        Args:
-            graph(IrGraph): the applied graph.
-        """
-
         assert isinstance(graph,
                           IrGraph), 'graph must be the instance of IrGraph.'
-        ops = graph.all_op_nodes()
 
-        persistable_vars = [p.name() for p in graph.all_persistable_nodes()]
-        # Collect the InScales and max_range to calculate the new scales for MKL-DNN 
-        # INT8 conv2d and mul
-        for op_node in ops:
-            if op_node.name() in self.dequantize_type:
-                input_name = op_node.input("X")[0]
-                scale_name = op_node.input("Scale")[0]
-                self.InScale[input_name] = self._load_param(self._scope,
-                                                            scale_name)[0]
-                self.max_range[input_name] = op_node.op().attr("max_range")
-                self.new_output[input_name] = op_node.output("Out")[0]
+        graph = self._gather_scales(graph)
+        graph = self._remove_fake_ops(graph)
+        graph = self._update_pooling_scales(graph)
+        graph = self._dequantize_weights(graph)
+        graph = self._optimize_fp32_graph(graph)
+        graph = self._compute_weight_scales(graph)
+        graph = self._quantize_fp32_graph(graph)
+        graph = self._remove_unused_var_nodes(graph)
+        return graph
 
-        for op_node in ops:
-            if op_node.name() in self._quantizable_ops:
-                if op_node.name() in self._conv_ops:
-                    self._transform_to_conv_mkldnn(graph, op_node)
-                else:
-                    self._transform_to_mul_mkldnn(graph, op_node)
-            elif op_node.name() in self.quantize_type:
-                self._transform_to_quantize_mkldnn(graph, op_node)
-            elif op_node.name() in self.dequantize_type:
-                self._remove_fake_dequantize_op(graph, op_node)
-            self._remove_unused_var_nodes(graph)
+    def _convert_scale2tensor(self, scale):
+        tensor = core.LoDTensor()
+        tensor.set(scale, core.CPUPlace())
+        return tensor
+
+    def _gather_scales(self, graph):
+        for op in graph.all_op_nodes():
+            if op.name() in self.quantize_types:
+                bit_length = op.op().attr("bit_length")
+                assert bit_length == 8, 'Unsupported number quantization bits ({}). Only 8 is supported now.'.format(
+                    bit_length)
+
+                input_name = op.input("X")[0]
+                scale_name = op.input("InScale")[0]
+                scale = np.array(1.0 / self._load_param(
```

Comment:
Add comment here: Gather the new weights scale after folding batchnorm in convolution.